feat: Introduce udemy/import-blacklist

kevinzang opened the pull request:
##### *To:*
@cansin @udemy/team-f 

##### *What:*
feat: Introduce udemy/import-blacklist to prevent developers from writing certain imports.

##### *Trello:*
https://trello.com/c/O67N8QkA/1954-investigate-if-we-can-have-eslint-prevent-full-library-imports

##### *What did you test:*
I manually added import-blacklist/index.js to node_modules in website-django, and ran it against the codebase. It found a few bad imports for `'numeral'` :)

##### *What dashboards will you be monitoring:*
None

Commit:
feat: Introduce udemy/import-blacklist

kevinzang commented on packages/eslint-config-udemy-website/index.js:
```diff
@@ -10,7 +10,6 @@ module.exports = {
     plugins: [
         'udemy',
         'filenames',
-        'lodash',
```
I removed the lodash plugin introduced in https://github.com/udemy/js-tooling/pull/10 because the import-blacklist rule that I wrote is a generalized version of it.

BTW, what is the intended way to remove a dependency? There's no "lerna remove", as far as I know. I did `cd packages/eslint-config-udemy/website && yarn remove eslint-plugin-lodash`...

kevinzang replied:
Cool, I've documented this in https://github.com/udemy/js-tooling/blob/feature/import-blacklist/README.md.

kevinzang commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Investigate if we can have eslint prevent full library imports](https://trello.com/c/O67N8QkA/1954-investigate-if-we-can-have-eslint-prevent-full-library-imports)

kevinzang commented on packages/eslint-config-udemy-website/index.js:
```diff
@@ -21,9 +20,9 @@ module.exports = {
     rules: {
-        'filenames/match-regex': ['error', '^(?:[a-z0-9\\-]+(?:\\.(?:jqui-widget|ng-(?:constant|controller|directive|factory|filter|provider|service)|react-(?:component|proptypes)|mobx-(?:model|store)))?(?:\\.spec)?)$'],
-        'lodash/import-scope': ['error', 'method'],
+        'filenames/match-regex': ['error', '^(?:[a-z0-9\\-]+(?:\\.(?:ng-(?:constant|controller|directive|factory|filter|provider|service)|react-(?:component|proptypes)|mobx-(?:model|store)))?(?:\\.spec)?)$'],
```
https://trello.com/c/tKqRRBdE/1911-after-jqui-is-deleted-remove-jqui-widget-from-eslint-config-udemy-website-filenames-match-regex

kevinzang commented:
![](https://github.trello.services/images/mini-trello-icon.png) [After jqui is deleted, remove jqui-widget from eslint-config-udemy-website filenames/match-regex](https://trello.com/c/tKqRRBdE/1911-after-jqui-is-deleted-remove-jqui-widget-from-eslint-config-udemy-website-filenames-match-regex)

inancsevinc commented on packages/eslint-plugin-udemy/rules/import-blacklist/index.js:
```diff
@@ -0,0 +1,73 @@
+    {
+        // For improved JS bundling
+        pattern: '^lodash(?:\\.js)?$',
+        message: 'Please import from lodash/lib/, not from lodash directly',
```
@kevinzang maybe tweak the message here? We allow things like `import mapValues from 'lodash/mapValues'`.

kevinzang replied:
Thanks! I've updated the messages.

cansin reviewed (approved):
1 comment, o/w LGTM!

jjinux commented on packages/eslint-plugin-udemy/rules/import-blacklist/README.md:
```diff
@@ -0,0 +1,50 @@
+
+`udemy/import-blacklist` blacklists certain imports in JS files. For example, it might tell you to `please import from foo/lib/, not from foo directly`. There are several reasons we enforce this rule:
+
+- Sometimes, a more specific import, e.g. `import bar from 'foo/lib/bar';`, can significantly reduce JS bundle size compared to e.g. `import { bar } from 'foo';`, since `'foo'` may contain many modules that the importing file is not actually using.
```
@kngroo ^

jjinux commented on packages/eslint-plugin-udemy/rules/import-blacklist/index.js:
```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+const blacklist = [
```
👍

jjinux replied:
Maybe you can provide default configuration that they can embrace and extend or replace entirely if they want to.

cansin replied:
I feel like the current blacklist can't be a sane default though since it's quite unique to website-django/static use case (e.g. especially the exception paths and the set of libraries used together).

kevinzang replied:
Thanks! I've moved the blacklist to a context option.

jjinux commented on packages/eslint-plugin-udemy/rules/import-blacklist/index.js:
```diff
@@ -0,0 +1,73 @@
+    },
+    {
+        // For code consistency
+        pattern: '^mobx/lib/mobx(?:\\.js)?$',
```
Thank you :) It irritates me when people do that.

Commit:
refactor: Move blacklist to context option

Commit:
docs: Document how to remove a dependency

Commit:
test: simplify the test cases

Commit:
Publish

 - babel-polyfill-udemy-website@0.6.4
 - babel-preset-udemy-website@0.6.4
 - eslint-config-udemy-basics@3.1.17
 - eslint-config-udemy-website@4.1.0
 - eslint-plugin-udemy@4.1.0

cansin commented:
I realized we forgot to list this new rule at `https://github.com/udemy/js-tooling/blob/master/packages/eslint-plugin-udemy/README.md`.

kevinzang commented:
@cansin my bad. I've created https://github.com/udemy/js-tooling/pull/21 to update that doc.
